runtime/src/tiered_storage.rs

```diff
@@ -8,6 +10,65 @@ pub mod meta;
 pub mod mmap_utils;
 pub mod readable;
 
-use crate::tiered_storage::error::TieredStorageError;
+use {
+    error::TieredStorageError,
+    footer::{AccountBlockFormat, AccountMetaFormat, OwnersBlockFormat},
+    index::AccountIndexFormat,
+    std::path::{Path, PathBuf},
+};
 
 pub type TieredStorageResult<T> = Result<T, TieredStorageError>;
+
+/// The struct that defines the formats of all building blocks of a
+/// TieredAccountsFile.
+#[derive(Debug)]
+pub struct TieredAccountsFileFormat {
+    pub meta_entry_size: usize,
+    pub account_meta_format: AccountMetaFormat,
+    pub owners_block_format: OwnersBlockFormat,
+    pub account_index_format: AccountIndexFormat,
+    pub account_block_format: AccountBlockFormat,
+}
+
+/// The struct for a TieredAccountsFile.
+#[derive(Debug)]
+pub struct TieredAccountsFile {
+    format: Option<&'static TieredAccountsFileFormat>,
```

Comment:
It is optional because it is only used in the write-path (in the read path, we have it under the footer).
Owning outright should be a better option, it also allows us to take a validator argument to further customized it if we want to enable such behavior.

Will change it to instance type.

Reply:
Forgot to address this one.  Will do.

Reply:
This one struct contains four enums.  Not sure if an enum is the best here.

Reply:
The PR currently defines HOT_FORMAT as the default format for the hot storage.  Let me know if you want it to be an enum that contains four formats instead.

Reply:
Oh oh, then I think it needs to be a struct because new_writable will become read-only after it finishes append_accounts().  accounts_db will not close it and reopen it.  It instead directly takes the written file.

Reply:
So what accounts db does when it wants to append accounts are:

* Creates an AcountStorageEntry
* During the creation of AccountsStorageEntry, it also creates an accounts file (in our case, we will invoke new_writable).
* invoke append_accounts (in our case, this will turn the file to be readonly after the call).
* Return the AccountStorageEntry.

So it has to be the same accounts file instance that is associated with the AccountStorageEntry.  If we use an enum that contains write and read, then we won't be able to do it with just one instance.

Reply:
> Gotcha. After the call to append_accounts(), what happens to the format field?

It remains untouched.  The thing that changed is that creates a reader instance inside a OnceCell.

Here:
https://github.com/solana-labs/solana/pull/32579#discussion_r1270973592

Reply:
Note that accounts_db requires any AccountsFile to not be `mut`.

Reply:
Somehow missed your last comment.  Having an enum field under the TieredStorage struct is an interesting idea.

Let me play around a little bit.

Reply:
The proposed enum method above seems to require the instance to be mut, while `fn append_accounts` takes `&self`.  If it requires mutex or something to workaround, I prefer the current approach in this PR that uses OnceCell as it matches the behavior that the reader can only be initialized once.

Can I know your opinion on this?

```
    fn append_accounts(...) {
        match self.0 {
            Reader(..) => unimplemented!(),
            Writer(path, format) => {
                let accounts_file = self.do_the_append_accounts();
                self.0 = TieredStorageKind::Reader(accounts_file);
            }
        }
    }
```

Reply:
Offline discussed with @brooksprumo.  `OnceCell` implementation looks better as it is lighter than regular `Cell`.

Reply:
`OnceCell` is used [here](https://github.com/solana-labs/solana/pull/32579#discussion_r1273854020) in #32579 (the third PR of this series).